In [1]:
import pandas as pd
import numpy as np

In [2]:
jan = pd.read_parquet(r"C:\Users\Ritu\Downloads\Parquet Data\yellow_tripdata_2026-01.parquet")
feb = pd.read_parquet(r"C:\Users\Ritu\Downloads\Parquet Data\yellow_tripdata_2026-02.parquet")
mar = pd.read_parquet(r"C:\Users\Ritu\Downloads\Parquet Data\yellow_tripdata_2026-03.parquet")
df = pd.concat([jan,feb,mar],ignore_index=True)

In [3]:
df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
0,2,2026-01-01 00:54:04,2026-01-01 00:59:37,1.0,0.97,1.0,N,239,238,1,7.2,1.00,0.5,3.66,0.0,1.0,15.86,2.5,0.0,0.00
1,1,2026-01-01 00:34:04,2026-01-01 00:39:47,0.0,0.90,1.0,N,163,162,2,7.9,4.25,0.5,0.00,0.0,1.0,13.65,2.5,0.0,0.75
2,1,2026-01-01 00:57:06,2026-01-01 01:05:59,0.0,1.40,1.0,N,43,237,1,10.7,4.25,0.5,2.50,0.0,1.0,18.95,2.5,0.0,0.75
3,2,2026-01-01 00:15:22,2026-01-01 00:58:10,4.0,5.58,1.0,N,142,209,1,38.7,1.00,0.5,11.11,0.0,1.0,55.56,2.5,0.0,0.75
4,2,2026-01-01 00:27:13,2026-01-01 00:40:43,0.0,2.16,1.0,N,88,144,1,13.5,1.00,0.5,3.85,0.0,1.0,23.10,2.5,0.0,0.75


In [4]:
df.shape

(11077206, 20)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11077206 entries, 0 to 11077205
Data columns (total 20 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int32         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int32         
 8   DOLocationID           int32         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  Airport_fee         

In [6]:
df.describe()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
count,1.107721e+07,11077206,11077206,8.020083e+06,1.107721e+07,8.020083e+06,1.107721e+07,1.107721e+07,1.107721e+07,1.107721e+07,1.107721e+07,1.107721e+07,1.107721e+07,1.107721e+07,1.107721e+07,1.107721e+07,8.020083e+06,8.020083e+06,1.107721e+07
mean,1.876552e+00,2026-02-15 16:11:28.931824,2026-02-15 16:28:54.533629,1.238929e+00,6.221664e+00,5.384528e+00,1.613870e+02,1.608872e+02,8.508780e-01,2.124950e+01,1.053296e+00,4.862019e-01,2.706643e+00,5.067500e-01,9.509915e-01,2.979328e+01,2.172388e+00,1.494923e-01,5.241209e-01
min,1.000000e+00,2008-12-31 23:03:20,2009-01-01 00:07:36,0.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,-2.555200e+03,-7.500000e+00,-5.000000e-01,-8.888000e+01,-1.117900e+02,-1.000000e+00,-2.560200e+03,-2.500000e+00,-2.000000e+00,-7.500000e-01
25%,2.000000e+00,2026-01-24 00:02:16,2026-01-24 00:16:40,1.000000e+00,1.000000e+00,1.000000e+00,1.140000e+02,1.070000e+02,0.000000e+00,1.000000e+01,0.000000e+00,5.000000e-01,0.000000e+00,0.000000e+00,1.000000e+00,1.725000e+01,2.500000e+00,0.000000e+00,0.000000e+00
50%,2.000000e+00,2026-02-14 21:54:05,2026-02-14 22:12:57,1.000000e+00,1.800000e+00,1.000000e+00,1.610000e+02,1.620000e+02,1.000000e+00,1.560000e+01,0.000000e+00,5.000000e-01,2.000000e+00,0.000000e+00,1.000000e+00,2.355000e+01,2.500000e+00,0.000000e+00,7.500000e-01
75%,2.000000e+00,2026-03-10 14:52:12,2026-03-10 15:13:28,1.000000e+00,3.700000e+00,1.000000e+00,2.330000e+02,2.340000e+02,1.000000e+00,2.680000e+01,2.500000e+00,5.000000e-01,3.850000e+00,0.000000e+00,1.000000e+00,3.471000e+01,2.500000e+00,0.000000e+00,7.500000e-01
max,7.000000e+00,2026-04-01 00:06:25,2026-04-02 16:03:47,9.000000e+00,3.285222e+05,9.900000e+01,2.650000e+02,2.650000e+02,4.000000e+00,2.555200e+03,2.071000e+01,4.750000e+00,7.660000e+02,1.400000e+03,1.000000e+00,2.560200e+03,2.500000e+00,2.700000e+01,7.500000e-01
std,7.105294e-01,NaN,NaN,6.461159e-01,6.207150e+02,2.003832e+01,6.690951e+01,7.093040e+01,6.758964e-01,1.873519e+01,1.718890e+00,1.025176e-01,3.892603e+00,2.179020e+00,2.492188e-01,2.237361e+01,9.048935e-01,5.378711e-01,3.523499e-01


In [7]:
df.isnull().sum()

VendorID                       0
tpep_pickup_datetime           0
tpep_dropoff_datetime          0
passenger_count          3057123
trip_distance                  0
RatecodeID               3057123
store_and_fwd_flag       3057123
PULocationID                   0
DOLocationID                   0
payment_type                   0
fare_amount                    0
extra                          0
mta_tax                        0
tip_amount                     0
tolls_amount                   0
improvement_surcharge          0
total_amount                   0
congestion_surcharge     3057123
Airport_fee              3057123
cbd_congestion_fee             0
dtype: int64

In [8]:
(df['trip_distance'] <= 0).sum()

np.int64(370198)

In [9]:
(df['fare_amount'] <= 0).sum()

np.int64(94364)

In [10]:
(df['total_amount'] <= 0).sum()

np.int64(90080)

In [11]:
(df['passenger_count'] <= 0).sum()

np.int64(40638)

In [12]:
df[df['trip_distance'] <= 0][['trip_distance','fare_amount','total_amount']].head()

,trip_distance,fare_amount,total_amount
118,0.0,3.0,8.00
329,0.0,18.4,24.15
330,0.0,12.8,23.15
360,0.0,50.0,54.25
634,0.0,50.0,54.25


In [13]:
df['trip_distance'].describe()

count    1.107721e+07
mean     6.221664e+00
std      6.207150e+02
min      0.000000e+00
25%      1.000000e+00
50%      1.800000e+00
75%      3.700000e+00
max      3.285222e+05
Name: trip_distance, dtype: float64

In [14]:
df[df['trip_distance'] > 100][
    ['trip_distance','fare_amount','total_amount']
].sort_values('trip_distance', ascending=False).head(20)

,trip_distance,fare_amount,total_amount
6826279,328522.20,31.11,35.86
7058668,298292.14,74.10,75.60
10403469,288381.68,21.39,31.37
3524923,269097.48,12.19,16.94
6822975,240830.74,15.07,20.57
6187804,237905.70,44.19,48.94
3632367,237567.19,72.99,81.95
3285750,236216.69,13.58,21.33
3131676,236183.06,22.75,34.96
2704619,221544.91,31.21,39.55


In [15]:
(df['trip_distance'] > 100).sum()

np.int64(495)

I found 495 records with unusually high trip distances (>100 miles). Since they fall outside the normal distribution of NYC taxi trips, I treated them as outliers and removed them to maintain data quality.

In [16]:
df_clean = df[df['trip_distance'] <= 100]
df_clean

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
0,2,2026-01-01 00:54:04,2026-01-01 00:59:37,1.0,0.97,1.0,N,239,238,1,7.20,1.00,0.5,3.66,0.0,1.0,15.86,2.5,0.0,0.00
1,1,2026-01-01 00:34:04,2026-01-01 00:39:47,0.0,0.90,1.0,N,163,162,2,7.90,4.25,0.5,0.00,0.0,1.0,13.65,2.5,0.0,0.75
2,1,2026-01-01 00:57:06,2026-01-01 01:05:59,0.0,1.40,1.0,N,43,237,1,10.70,4.25,0.5,2.50,0.0,1.0,18.95,2.5,0.0,0.75
3,2,2026-01-01 00:15:22,2026-01-01 00:58:10,4.0,5.58,1.0,N,142,209,1,38.70,1.00,0.5,11.11,0.0,1.0,55.56,2.5,0.0,0.75
4,2,2026-01-01 00:27:13,2026-01-01 00:40:43,0.0,2.16,1.0,N,88,144,1,13.50,1.00,0.5,3.85,0.0,1.0,23.10,2.5,0.0,0.75
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11077201,2,2026-03-31 23:10:43,2026-03-31 23:33:02,NaN,8.47,NaN,None,13,263,0,50.14,0.00,0.5,0.00,0.0,1.0,60.37,NaN,NaN,0.75
11077202,2,2026-03-31 23:41:22,2026-03-31 23:52:36,NaN,1.99,NaN,None,24,42,0,12.32,0.00,0.5,0.00,0.0,1.0,13.82,NaN,NaN,0.00
11077203,2,2026-03-31 23:39:38,2026-04-01 00:13:44,NaN,10.21,NaN,None,62,237,0,47.95,0.00,0.5,0.00,0.0,1.0,52.70,NaN,NaN,0.75
11077204,2,2026-03-31 23:57:57,2026-04-01 00:29:33,NaN,7.87,NaN,None,61,141,0,49.93,0.00,0.5,0.00,0.0,1.0,53.93,NaN,NaN,0.00


In [17]:
df[['fare_amount']].describe()

,fare_amount
count,1.107721e+07
mean,2.124950e+01
std,1.873519e+01
min,-2.555200e+03
25%,1.000000e+01
50%,1.560000e+01
75%,2.680000e+01
max,2.555200e+03


In [18]:
df[df['fare_amount'] < 0][['trip_distance','fare_amount','total_amount']].head(10)

,trip_distance,fare_amount,total_amount
74,0.63,-5.1,-10.10
248,1.92,-22.6,-28.35
368,0.02,-3.0,-8.75
428,0.03,-3.0,-8.75
649,3.12,-22.6,-28.35
871,1.03,-8.6,-13.60
980,0.01,-70.0,-71.50
1046,0.14,-3.7,-9.45
1076,3.63,-17.7,-22.70
1106,0.01,-3.0,-8.00


In [19]:
df_clean = df[df['fare_amount'] >= 0].copy()
df_clean

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
0,2,2026-01-01 00:54:04,2026-01-01 00:59:37,1.0,0.97,1.0,N,239,238,1,7.20,1.00,0.5,3.66,0.0,1.0,15.86,2.5,0.0,0.00
1,1,2026-01-01 00:34:04,2026-01-01 00:39:47,0.0,0.90,1.0,N,163,162,2,7.90,4.25,0.5,0.00,0.0,1.0,13.65,2.5,0.0,0.75
2,1,2026-01-01 00:57:06,2026-01-01 01:05:59,0.0,1.40,1.0,N,43,237,1,10.70,4.25,0.5,2.50,0.0,1.0,18.95,2.5,0.0,0.75
3,2,2026-01-01 00:15:22,2026-01-01 00:58:10,4.0,5.58,1.0,N,142,209,1,38.70,1.00,0.5,11.11,0.0,1.0,55.56,2.5,0.0,0.75
4,2,2026-01-01 00:27:13,2026-01-01 00:40:43,0.0,2.16,1.0,N,88,144,1,13.50,1.00,0.5,3.85,0.0,1.0,23.10,2.5,0.0,0.75
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11077201,2,2026-03-31 23:10:43,2026-03-31 23:33:02,NaN,8.47,NaN,None,13,263,0,50.14,0.00,0.5,0.00,0.0,1.0,60.37,NaN,NaN,0.75
11077202,2,2026-03-31 23:41:22,2026-03-31 23:52:36,NaN,1.99,NaN,None,24,42,0,12.32,0.00,0.5,0.00,0.0,1.0,13.82,NaN,NaN,0.00
11077203,2,2026-03-31 23:39:38,2026-04-01 00:13:44,NaN,10.21,NaN,None,62,237,0,47.95,0.00,0.5,0.00,0.0,1.0,52.70,NaN,NaN,0.75
11077204,2,2026-03-31 23:57:57,2026-04-01 00:29:33,NaN,7.87,NaN,None,61,141,0,49.93,0.00,0.5,0.00,0.0,1.0,53.93,NaN,NaN,0.00


In [20]:
(df_clean['fare_amount'] < 0).sum()

np.int64(0)

In [21]:
df_clean = df_clean[df_clean['passenger_count'] > 0].copy()
df_clean

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
0,2,2026-01-01 00:54:04,2026-01-01 00:59:37,1.0,0.97,1.0,N,239,238,1,7.2,1.00,0.5,3.66,0.0,1.0,15.86,2.5,0.0,0.00
3,2,2026-01-01 00:15:22,2026-01-01 00:58:10,4.0,5.58,1.0,N,142,209,1,38.7,1.00,0.5,11.11,0.0,1.0,55.56,2.5,0.0,0.75
5,2,2026-01-01 00:47:11,2026-01-01 01:00:47,2.0,2.33,1.0,N,144,137,1,14.2,1.00,0.5,4.99,0.0,1.0,24.94,2.5,0.0,0.75
6,1,2026-01-01 00:17:54,2026-01-01 00:28:32,1.0,1.30,1.0,N,142,50,2,11.4,4.25,0.5,0.00,0.0,1.0,17.15,2.5,0.0,0.75
8,2,2026-01-01 00:34:14,2026-01-01 01:11:58,1.0,5.34,1.0,N,161,45,1,37.3,1.00,0.5,8.61,0.0,1.0,51.66,2.5,0.0,0.75
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10131453,2,2026-03-31 23:17:20,2026-03-31 23:27:29,1.0,2.29,1.0,N,236,163,1,12.8,1.00,0.5,4.64,0.0,1.0,23.19,2.5,0.0,0.75
10131454,2,2026-03-31 23:30:28,2026-03-31 23:40:04,1.0,1.90,1.0,N,163,246,1,11.4,1.00,0.5,3.43,0.0,1.0,20.58,2.5,0.0,0.75
10131455,2,2026-03-31 23:15:21,2026-03-31 23:30:37,1.0,2.62,1.0,N,142,107,1,16.3,1.00,0.5,2.22,0.0,1.0,24.27,2.5,0.0,0.75
10131456,2,2026-03-31 23:02:01,2026-03-31 23:06:57,1.0,1.03,1.0,N,170,100,1,7.2,1.00,0.5,0.00,0.0,1.0,12.95,2.5,0.0,0.75


In [22]:
(df_clean['passenger_count'] <= 0).sum()

np.int64(0)

In [23]:
(df_clean['total_amount'] < 0).sum()

np.int64(1710)

In [24]:
df_clean = df_clean[df_clean["total_amount"] >= 0].copy()

In [25]:
df_clean['total_amount'].describe()

count    7.890952e+06
mean     2.920439e+01
std      2.315945e+01
min      0.000000e+00
25%      1.645000e+01
50%      2.155000e+01
75%      3.115000e+01
max      2.560200e+03
Name: total_amount, dtype: float64

In [26]:
(df_clean['fare_amount'] <= 0).sum()

np.int64(3227)

In [27]:
df_clean = df_clean[df_clean['fare_amount'] > 0].copy()

In [28]:
(df_clean['fare_amount'] <= 0).sum()

np.int64(0)

In [29]:
df_clean.describe()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
count,7.887725e+06,7887725,7887725,7.887725e+06,7.887725e+06,7.887725e+06,7.887725e+06,7.887725e+06,7.887725e+06,7.887725e+06,7.887725e+06,7.887725e+06,7.887725e+06,7.887725e+06,7.887725e+06,7.887725e+06,7.887725e+06,7.887725e+06,7.887725e+06
mean,1.869869e+00,2026-02-16 11:58:00.538951,2026-02-16 12:15:16.331498,1.244514e+00,3.407785e+00,5.451453e+00,1.650768e+02,1.647603e+02,1.149734e+00,1.969277e+01,1.465949e+00,4.921011e-01,3.625223e+00,5.806748e-01,9.555485e-01,2.921534e+01,2.218350e+00,1.557828e-01,5.278798e-01
min,1.000000e+00,2008-12-31 23:03:20,2009-01-01 00:07:36,1.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e-02,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,7.000000e-01,0.000000e+00,0.000000e+00,0.000000e+00
25%,2.000000e+00,2026-01-23 20:43:33,2026-01-23 20:58:09,1.000000e+00,9.700000e-01,1.000000e+00,1.320000e+02,1.140000e+02,1.000000e+00,9.300000e+00,0.000000e+00,5.000000e-01,1.000000e+00,0.000000e+00,1.000000e+00,1.645000e+01,2.500000e+00,0.000000e+00,0.000000e+00
50%,2.000000e+00,2026-02-16 21:51:02,2026-02-16 22:05:34,1.000000e+00,1.660000e+00,1.000000e+00,1.620000e+02,1.620000e+02,1.000000e+00,1.350000e+01,1.000000e+00,5.000000e-01,2.950000e+00,0.000000e+00,1.000000e+00,2.155000e+01,2.500000e+00,0.000000e+00,7.500000e-01
75%,2.000000e+00,2026-03-11 17:53:46,2026-03-11 18:12:32,1.000000e+00,3.300000e+00,1.000000e+00,2.340000e+02,2.340000e+02,1.000000e+00,2.190000e+01,2.500000e+00,5.000000e-01,4.480000e+00,0.000000e+00,1.000000e+00,3.115000e+01,2.500000e+00,0.000000e+00,7.500000e-01
max,7.000000e+00,2026-04-01 00:06:25,2026-04-02 16:03:47,9.000000e+00,9.877358e+04,9.900000e+01,2.650000e+02,2.650000e+02,4.000000e+00,2.555200e+03,2.071000e+01,4.750000e+00,7.660000e+02,1.400000e+03,1.000000e+00,2.560200e+03,2.500000e+00,2.700000e+01,7.500000e-01
std,7.885865e-01,NaN,NaN,6.410446e-01,3.990165e+01,2.019062e+01,6.370524e+01,6.959770e+01,4.329675e-01,1.884418e+01,1.848079e+00,6.246187e-02,4.141993e+00,2.321905e+00,2.060960e-01,2.315772e+01,7.904423e-01,5.335433e-01,3.424219e-01


In [30]:
df_clean = df_clean[(df_clean['trip_distance'] <= 100) &(df_clean['fare_amount'] > 0)].copy()
df_clean

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
0,2,2026-01-01 00:54:04,2026-01-01 00:59:37,1.0,0.97,1.0,N,239,238,1,7.2,1.00,0.5,3.66,0.0,1.0,15.86,2.5,0.0,0.00
3,2,2026-01-01 00:15:22,2026-01-01 00:58:10,4.0,5.58,1.0,N,142,209,1,38.7,1.00,0.5,11.11,0.0,1.0,55.56,2.5,0.0,0.75
5,2,2026-01-01 00:47:11,2026-01-01 01:00:47,2.0,2.33,1.0,N,144,137,1,14.2,1.00,0.5,4.99,0.0,1.0,24.94,2.5,0.0,0.75
6,1,2026-01-01 00:17:54,2026-01-01 00:28:32,1.0,1.30,1.0,N,142,50,2,11.4,4.25,0.5,0.00,0.0,1.0,17.15,2.5,0.0,0.75
8,2,2026-01-01 00:34:14,2026-01-01 01:11:58,1.0,5.34,1.0,N,161,45,1,37.3,1.00,0.5,8.61,0.0,1.0,51.66,2.5,0.0,0.75
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10131453,2,2026-03-31 23:17:20,2026-03-31 23:27:29,1.0,2.29,1.0,N,236,163,1,12.8,1.00,0.5,4.64,0.0,1.0,23.19,2.5,0.0,0.75
10131454,2,2026-03-31 23:30:28,2026-03-31 23:40:04,1.0,1.90,1.0,N,163,246,1,11.4,1.00,0.5,3.43,0.0,1.0,20.58,2.5,0.0,0.75
10131455,2,2026-03-31 23:15:21,2026-03-31 23:30:37,1.0,2.62,1.0,N,142,107,1,16.3,1.00,0.5,2.22,0.0,1.0,24.27,2.5,0.0,0.75
10131456,2,2026-03-31 23:02:01,2026-03-31 23:06:57,1.0,1.03,1.0,N,170,100,1,7.2,1.00,0.5,0.00,0.0,1.0,12.95,2.5,0.0,0.75


In [31]:
print(df_clean.shape)

(7887599, 20)


In [32]:
df_clean['trip_duration_minutes'] = (df_clean['tpep_dropoff_datetime'] - df_clean['tpep_pickup_datetime']).dt.total_seconds() / 60
df_clean['trip_duration_minutes']

0            5.550000
3           42.800000
5           13.600000
6           10.633333
8           37.733333
              ...    
10131453    10.150000
10131454     9.600000
10131455    15.266667
10131456     4.933333
10131457     6.383333
Name: trip_duration_minutes, Length: 7887599, dtype: float64

In [33]:
df_clean['trip_duration_minutes'].describe()

count    7.887599e+06
mean     1.726132e+01
std      2.840406e+01
min     -1.170000e+01
25%      7.333333e+00
50%      1.240000e+01
75%      2.078333e+01
max      9.265150e+03
Name: trip_duration_minutes, dtype: float64

In [34]:
(df_clean['trip_duration_minutes'] <= 0).sum()

np.int64(134350)

In [35]:
df_clean = df_clean[df_clean['trip_duration_minutes'] > 0].copy()

In [36]:
(df_clean['trip_duration_minutes'] <= 0).sum()

np.int64(0)

In [37]:
df_clean[df_clean['trip_duration_minutes'] > 300][['trip_duration_minutes', 'trip_distance', 'fare_amount']
].sort_values('trip_duration_minutes', ascending=False).head(20)

,trip_duration_minutes,trip_distance,fare_amount
5420879,9265.150000,0.00,4.40
8805594,7620.600000,0.00,3.00
1560244,7507.900000,0.00,70.00
7524635,7482.066667,0.07,21.20
9606280,7435.666667,2.47,45.70
9288806,6746.933333,18.50,70.20
572461,5954.233333,0.00,3.70
710966,5293.333333,2.90,15.60
8595917,5291.766667,15.93,70.00
2027357,5154.583333,4.86,67.40


In [38]:
df_clean[(df_clean['trip_duration_minutes'] > 300) &(df_clean['trip_distance'] == 0)].shape

(289, 21)

In [39]:
(df_clean['trip_duration_minutes'] > 120).sum()

np.int64(12493)

In [40]:
(df_clean['trip_duration_minutes'] > 180).sum()

np.int64(4255)

In [41]:
(df_clean['trip_duration_minutes'] > 300).sum()

np.int64(3410)

In [42]:
df_clean['trip_duration_minutes'].describe(percentiles=[0.90, 0.95, 0.99, 0.995, 0.999])

count    7.753249e+06
mean     1.756042e+01
std      2.855728e+01
min      1.666667e-02
50%      1.261667e+01
90%      3.476667e+01
95%      4.795000e+01
99%      7.901667e+01
99.5%    9.318333e+01
99.9%    1.346625e+02
max      9.265150e+03
Name: trip_duration_minutes, dtype: float64

In [46]:
df_clean = df_clean[
    (df_clean["trip_duration_minutes"] <= 180) &
    (df_clean["trip_distance"] > 0)
]

In [47]:
df_clean = df_clean[df_clean["trip_distance"] <= 50].copy()

In [48]:
df_clean['trip_duration_minutes'].max()

179.96666666666667

In [49]:
print(df_clean.shape)

(7663542, 21)


In [50]:
df_clean.describe()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee,trip_duration_minutes
count,7.663542e+06,7663542,7663542,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06
mean,1.781883e+00,2026-02-16 12:27:10.153777,2026-02-16 12:44:24.700671,1.243473e+00,3.421717e+00,5.524068e+00,1.650353e+02,1.647328e+02,1.144538e+00,1.953877e+01,1.497266e+00,4.940286e-01,3.615113e+00,5.837416e-01,9.546211e-01,2.906858e+01,2.224690e+00,1.530083e-01,5.298751e-01,1.724245e+01
min,1.000000e+00,2009-01-01 00:02:17,2009-01-01 00:07:36,1.000000e+00,1.000000e-02,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e-02,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,7.000000e-01,0.000000e+00,0.000000e+00,0.000000e+00,1.666667e-02
25%,2.000000e+00,2026-01-23 20:42:28,2026-01-23 20:56:35.500000,1.000000e+00,1.000000e+00,1.000000e+00,1.320000e+02,1.140000e+02,1.000000e+00,9.300000e+00,0.000000e+00,5.000000e-01,1.000000e+00,0.000000e+00,1.000000e+00,1.650000e+01,2.500000e+00,0.000000e+00,0.000000e+00,7.683333e+00
50%,2.000000e+00,2026-02-16 23:48:13,2026-02-17 00:04:10.500000,1.000000e+00,1.690000e+00,1.000000e+00,1.620000e+02,1.620000e+02,1.000000e+00,1.350000e+01,1.000000e+00,5.000000e-01,2.950000e+00,0.000000e+00,1.000000e+00,2.160000e+01,2.500000e+00,0.000000e+00,7.500000e-01,1.270000e+01
75%,2.000000e+00,2026-03-11 18:08:41.750000,2026-03-11 18:26:57,1.000000e+00,3.340000e+00,1.000000e+00,2.340000e+02,2.340000e+02,1.000000e+00,2.190000e+01,2.500000e+00,5.000000e-01,4.480000e+00,0.000000e+00,1.000000e+00,3.114000e+01,2.500000e+00,0.000000e+00,7.500000e-01,2.110000e+01
max,2.000000e+00,2026-04-01 00:06:25,2026-04-01 01:22:42,8.000000e+00,5.000000e+01,9.900000e+01,2.650000e+02,2.650000e+02,4.000000e+00,9.500000e+02,2.071000e+01,4.750000e+00,7.660000e+02,1.222200e+02,1.000000e+00,9.510000e+02,2.500000e+00,5.000000e+00,7.500000e-01,1.799667e+02
std,4.129673e-01,NaN,NaN,6.393259e-01,4.513645e+00,2.038953e+01,6.366390e+01,6.961333e+01,4.220749e-01,1.769545e+01,1.855627e+00,5.445070e-02,4.024379e+00,2.260882e+00,2.081337e-01,2.212530e+01,7.826104e-01,5.027686e-01,3.415241e-01,1.529857e+01


In [51]:
df_clean[df_clean["tip_amount"] > 100].shape

(27, 21)

In [52]:
df_clean["tip_amount"] = df_clean["tip_amount"].clip(upper=100)

In [53]:
df_clean[df_clean['RatecodeID'].isin([1, 2, 3, 4, 5, 6])]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,...,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee,trip_duration_minutes
0,2,2026-01-01 00:54:04,2026-01-01 00:59:37,1.0,0.97,1.0,N,239,238,1,...,1.00,0.5,3.66,0.0,1.0,15.86,2.5,0.0,0.00,5.550000
3,2,2026-01-01 00:15:22,2026-01-01 00:58:10,4.0,5.58,1.0,N,142,209,1,...,1.00,0.5,11.11,0.0,1.0,55.56,2.5,0.0,0.75,42.800000
5,2,2026-01-01 00:47:11,2026-01-01 01:00:47,2.0,2.33,1.0,N,144,137,1,...,1.00,0.5,4.99,0.0,1.0,24.94,2.5,0.0,0.75,13.600000
6,1,2026-01-01 00:17:54,2026-01-01 00:28:32,1.0,1.30,1.0,N,142,50,2,...,4.25,0.5,0.00,0.0,1.0,17.15,2.5,0.0,0.75,10.633333
8,2,2026-01-01 00:34:14,2026-01-01 01:11:58,1.0,5.34,1.0,N,161,45,1,...,1.00,0.5,8.61,0.0,1.0,51.66,2.5,0.0,0.75,37.733333
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10131453,2,2026-03-31 23:17:20,2026-03-31 23:27:29,1.0,2.29,1.0,N,236,163,1,...,1.00,0.5,4.64,0.0,1.0,23.19,2.5,0.0,0.75,10.150000
10131454,2,2026-03-31 23:30:28,2026-03-31 23:40:04,1.0,1.90,1.0,N,163,246,1,...,1.00,0.5,3.43,0.0,1.0,20.58,2.5,0.0,0.75,9.600000
10131455,2,2026-03-31 23:15:21,2026-03-31 23:30:37,1.0,2.62,1.0,N,142,107,1,...,1.00,0.5,2.22,0.0,1.0,24.27,2.5,0.0,0.75,15.266667
10131456,2,2026-03-31 23:02:01,2026-03-31 23:06:57,1.0,1.03,1.0,N,170,100,1,...,1.00,0.5,0.00,0.0,1.0,12.95,2.5,0.0,0.75,4.933333


In [54]:
df_clean[~df_clean['RatecodeID'].isin([1, 2, 3, 4, 5, 6])].shape

(347920, 21)

In [55]:
df_clean[
    (~df_clean['RatecodeID'].isin([1,2,3,4,5,6])) &
    (df_clean['RatecodeID'].notna())
]['RatecodeID'].value_counts()

RatecodeID
99.0    347920
Name: count, dtype: int64

In [56]:
# fill missing
df_clean['RatecodeID'] = df_clean['RatecodeID'].fillna(df_clean['RatecodeID'].mode()[0]).copy()

In [57]:
# fix invalid
mode_val = df_clean["RatecodeID"].mode()[0]
df_clean.loc[df_clean["RatecodeID"] > 6, "RatecodeID"] = mode_val

In [58]:
df_clean[
    (~df_clean['RatecodeID'].isin([1,2,3,4,5,6])) &
    (df_clean['RatecodeID'].notna())
]['RatecodeID'].value_counts()

Series([], Name: count, dtype: int64)

In [59]:
df_clean.agg(['min', 'max'])

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,...,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee,trip_duration_minutes
min,1,2009-01-01 00:02:17,2009-01-01 00:07:36,1.0,0.01,1.0,N,1,1,1,...,0.00,0.00,0.0,0.00,0.0,0.7,0.0,0.0,0.00,0.016667
max,2,2026-04-01 00:06:25,2026-04-01 01:22:42,8.0,50.00,6.0,Y,265,265,4,...,20.71,4.75,100.0,122.22,1.0,951.0,2.5,5.0,0.75,179.966667


In [60]:
df_clean.to_parquet(
    "nyc_taxi_cleaned_final.parquet",
    index=False
)
df_clean

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,...,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee,trip_duration_minutes
0,2,2026-01-01 00:54:04,2026-01-01 00:59:37,1.0,0.97,1.0,N,239,238,1,...,1.00,0.5,3.66,0.0,1.0,15.86,2.5,0.0,0.00,5.550000
3,2,2026-01-01 00:15:22,2026-01-01 00:58:10,4.0,5.58,1.0,N,142,209,1,...,1.00,0.5,11.11,0.0,1.0,55.56,2.5,0.0,0.75,42.800000
5,2,2026-01-01 00:47:11,2026-01-01 01:00:47,2.0,2.33,1.0,N,144,137,1,...,1.00,0.5,4.99,0.0,1.0,24.94,2.5,0.0,0.75,13.600000
6,1,2026-01-01 00:17:54,2026-01-01 00:28:32,1.0,1.30,1.0,N,142,50,2,...,4.25,0.5,0.00,0.0,1.0,17.15,2.5,0.0,0.75,10.633333
8,2,2026-01-01 00:34:14,2026-01-01 01:11:58,1.0,5.34,1.0,N,161,45,1,...,1.00,0.5,8.61,0.0,1.0,51.66,2.5,0.0,0.75,37.733333
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10131453,2,2026-03-31 23:17:20,2026-03-31 23:27:29,1.0,2.29,1.0,N,236,163,1,...,1.00,0.5,4.64,0.0,1.0,23.19,2.5,0.0,0.75,10.150000
10131454,2,2026-03-31 23:30:28,2026-03-31 23:40:04,1.0,1.90,1.0,N,163,246,1,...,1.00,0.5,3.43,0.0,1.0,20.58,2.5,0.0,0.75,9.600000
10131455,2,2026-03-31 23:15:21,2026-03-31 23:30:37,1.0,2.62,1.0,N,142,107,1,...,1.00,0.5,2.22,0.0,1.0,24.27,2.5,0.0,0.75,15.266667
10131456,2,2026-03-31 23:02:01,2026-03-31 23:06:57,1.0,1.03,1.0,N,170,100,1,...,1.00,0.5,0.00,0.0,1.0,12.95,2.5,0.0,0.75,4.933333


In [61]:
import pandas as pd

test = pd.read_parquet("nyc_taxi_cleaned_final.parquet")
print(test.shape)

(7663542, 21)


In [62]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7663542 entries, 0 to 10131457
Data columns (total 21 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int32         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int32         
 8   DOLocationID           int32         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  Airport_fee            flo

In [63]:
numeric_cols = df_clean.select_dtypes(include='number').columns
categorical_cols = df_clean.select_dtypes(include=['object', 'category']).columns
datetime_cols = df_clean.select_dtypes(include='datetime').columns

In [64]:
for col in numeric_cols:
    df_clean[col] = pd.to_numeric(df_clean[col], errors='coerce').fillna(df_clean[col].median())
    

In [65]:
for col in categorical_cols:
    df_clean[col] = (df_clean[col].fillna("Unknown").astype(str).str.strip())

In [66]:
df_clean.head(50)

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,...,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee,trip_duration_minutes
0,2,2026-01-01 00:54:04,2026-01-01 00:59:37,1.0,0.97,1.0,N,239,238,1,...,1.00,0.5,3.66,0.00,1.0,15.86,2.5,0.00,0.00,5.550000
3,2,2026-01-01 00:15:22,2026-01-01 00:58:10,4.0,5.58,1.0,N,142,209,1,...,1.00,0.5,11.11,0.00,1.0,55.56,2.5,0.00,0.75,42.800000
5,2,2026-01-01 00:47:11,2026-01-01 01:00:47,2.0,2.33,1.0,N,144,137,1,...,1.00,0.5,4.99,0.00,1.0,24.94,2.5,0.00,0.75,13.600000
6,1,2026-01-01 00:17:54,2026-01-01 00:28:32,1.0,1.30,1.0,N,142,50,2,...,4.25,0.5,0.00,0.00,1.0,17.15,2.5,0.00,0.75,10.633333
8,2,2026-01-01 00:34:14,2026-01-01 01:11:58,1.0,5.34,1.0,N,161,45,1,...,1.00,0.5,8.61,0.00,1.0,51.66,2.5,0.00,0.75,37.733333
9,2,2026-01-01 00:41:07,2026-01-01 00:50:42,3.0,1.83,1.0,N,237,263,1,...,1.00,0.5,2.36,0.00,1.0,18.06,2.5,0.00,0.00,9.583333
10,2,2026-01-01 00:12:46,2026-01-01 00:48:58,1.0,3.09,1.0,N,161,144,1,...,1.00,0.5,7.07,0.00,1.0,42.42,2.5,0.00,0.75,36.200000
11,2,2026-01-01 00:50:40,2026-01-01 01:03:09,1.0,1.26,1.0,N,144,79,1,...,1.00,0.5,3.57,0.00,1.0,21.42,2.5,0.00,0.75,12.483333
12,1,2026-01-01 00:21:33,2026-01-01 00:49:14,2.0,0.90,1.0,N,163,48,2,...,4.25,0.5,0.00,0.00,1.0,27.65,2.5,0.00,0.75,27.683333
13,2,2026-01-01 00:42:31,2026-01-01 00:44:28,3.0,0.33,1.0,N,234,90,2,...,1.00,0.5,0.00,0.00,1.0,10.15,2.5,0.00,0.75,1.950000


In [67]:
df_clean.describe()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee,trip_duration_minutes
count,7.663542e+06,7663542,7663542,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06,7.663542e+06
mean,1.781883e+00,2026-02-16 12:27:10.153777,2026-02-16 12:44:24.700671,1.243473e+00,3.421717e+00,1.074929e+00,1.650353e+02,1.647328e+02,1.144538e+00,1.953877e+01,1.497266e+00,4.940286e-01,3.614507e+00,5.837416e-01,9.546211e-01,2.906858e+01,2.224690e+00,1.530083e-01,5.298751e-01,1.724245e+01
min,1.000000e+00,2009-01-01 00:02:17,2009-01-01 00:07:36,1.000000e+00,1.000000e-02,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e-02,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,7.000000e-01,0.000000e+00,0.000000e+00,0.000000e+00,1.666667e-02
25%,2.000000e+00,2026-01-23 20:42:28,2026-01-23 20:56:35.500000,1.000000e+00,1.000000e+00,1.000000e+00,1.320000e+02,1.140000e+02,1.000000e+00,9.300000e+00,0.000000e+00,5.000000e-01,1.000000e+00,0.000000e+00,1.000000e+00,1.650000e+01,2.500000e+00,0.000000e+00,0.000000e+00,7.683333e+00
50%,2.000000e+00,2026-02-16 23:48:13,2026-02-17 00:04:10.500000,1.000000e+00,1.690000e+00,1.000000e+00,1.620000e+02,1.620000e+02,1.000000e+00,1.350000e+01,1.000000e+00,5.000000e-01,2.950000e+00,0.000000e+00,1.000000e+00,2.160000e+01,2.500000e+00,0.000000e+00,7.500000e-01,1.270000e+01
75%,2.000000e+00,2026-03-11 18:08:41.750000,2026-03-11 18:26:57,1.000000e+00,3.340000e+00,1.000000e+00,2.340000e+02,2.340000e+02,1.000000e+00,2.190000e+01,2.500000e+00,5.000000e-01,4.480000e+00,0.000000e+00,1.000000e+00,3.114000e+01,2.500000e+00,0.000000e+00,7.500000e-01,2.110000e+01
max,2.000000e+00,2026-04-01 00:06:25,2026-04-01 01:22:42,8.000000e+00,5.000000e+01,6.000000e+00,2.650000e+02,2.650000e+02,4.000000e+00,9.500000e+02,2.071000e+01,4.750000e+00,1.000000e+02,1.222200e+02,1.000000e+00,9.510000e+02,2.500000e+00,5.000000e+00,7.500000e-01,1.799667e+02
std,4.129673e-01,NaN,NaN,6.393259e-01,4.513645e+00,4.228965e-01,6.366390e+01,6.961333e+01,4.220749e-01,1.769545e+01,1.855627e+00,5.445070e-02,3.979330e+00,2.260882e+00,2.081337e-01,2.212530e+01,7.826104e-01,5.027686e-01,3.415241e-01,1.529857e+01


In [68]:
import os
os.getcwd()

'C:\\Users\\Ritu\\Downloads\\NYC'

In [ ]:
test.to_csv(
    r"C:\ProgramData\MySQL\MySQL Server 8.0\Uploads\cleaned_data_fixed.csv",
    index=False
)